src/compiler/checker.ts

```diff
@@ -4490,7 +4562,31 @@ namespace ts {
                 return Ternary.False;
             }
 
-            function unionTypeRelatedToType(source: UnionType, target: Type, reportErrors: boolean): Ternary {
+            function typeRelatedToEachType(source: Type, target: UnionOrIntersectionType, reportErrors: boolean): Ternary {
+                let result = Ternary.True;
+                let targetTypes = target.types;
+                for (let targetType of targetTypes) {
+                    let related = isRelatedTo(source, targetType, reportErrors);
+                    if (!related) {
+                        return Ternary.False;
+                    }
+                    result &= related;
+                }
+                return result;
+            }
+
+            function someTypeRelatedToType(source: UnionOrIntersectionType, target: Type, reportErrors: boolean): Ternary {
+                let sourceTypes = source.types;
+                for (let i = 0, len = sourceTypes.length; i < len; i++) {

```

Comment:
No, we need a regular for loop so we can compare the current index to `len - 1`.